### ISSUE TITLE
Zip Task reproducibleFileOrder does not enforce reproducible archives when "from" is called multiple times

### ISSUE THREAD

**Issue by atakar** [gradle/gradle#17248]
### Expected Behavior
The ouput of a ``Zip`` task with ``reproducibleFileOrder=true`` shall be the same regardless of the order of the `from` calls (*i.e.* it shall work in the same way as the ``UP-TO-DATE`` check works).

### Current Behavior
The ouput of a ``Zip`` task with ``reproducibleFileOrder=true`` changes when the order of the ``from`` calls varies.
The ``UP-TO-DATE`` checks however recognize the task as up-to-date regardless of the order of the ``from`` calls.

### Context
Note: The example project can be found as ``testProject.zip`` in the attachments [testProject.zip](https://github.com/gradle/gradle/files/6522016/testProject.zip).

Consider the following project structure, where ``file1`` to ``file4`` are empty files:

```
> testProject
    > build
    > src1
        > file1
        > file2
    > src2
        > file3
        > file4
    > build.gradle
```

Let us further consider that the ``build.gradle`` file declares the task ``testZip`` in the following way:
```
tasks.register("testZip", Zip).configure{ t ->
    t.from(project.files("src1"))
    t.from(project.files("src2"))
    t.archiveFileName = "Test.zip"
    t.destinationDirectory = project.buildDir
    t.preserveFileTimestamps = false
    t.reproducibleFileOrder = true
}
``` 

### Steps to Reproduce 
1. Run the task ``testZip``:
```console
PS C:\testProject> gradle.bat testZip

BUILD SUCCESSFUL in 1s
1 actionable task: 1 executed
```

2. Check and note the sha1sum of the output file. In my case, it gives ``71b78397``:
```console
PS C:\testProject> sha1sum.exe .\build\Test.zip
\71b783971908277e276efb79fc220b4310410250 *.\\build\\Test.zip
``` 

3. Change order of the ``from`` declarations in the ``testZip`` task to first ``src2`` then second ``src1`` in the ``build.gradle`` file and save:
```
tasks.register("testZip", Zip).configure{ t ->
    t.from(project.files("src2"))
    t.from(project.files("src1"))
//etc
}
```  

4. Run task ``testZip`` again. The task is ``UP-TO-DATE``, as it is expected to give the same output due to ``reproducibleFileOrder=true``:
```console
PS C:\testProject> gradle.bat testZip

BUILD SUCCESSFUL in 1s
1 actionable task: 1 up-to-date
```

5. However, the task will give another output, if we rerun the task with ``--rerun-tasks``. We can confirm that the sha1 hash has been changed to ``e225247b6``:
```console
PS C:\testProject> gradle.bat testZip --rerun-tasks

BUILD SUCCESSFUL in 1s
1 actionable task: 1 executed

PS C:\testProject> sha1sum.exe .\build\Test.zip

\e225247b6e1fe1ac1d71fda9a5ed78d649d48a3e *.\\build\\Test.zip

```
#### Gradle version
See output of command ``.\gradle.bat -v``:
```
------------------------------------------------------------
Gradle 7.0.2
------------------------------------------------------------

Build time:   2021-05-14 12:02:31 UTC
Revision:     1ef1b260d39daacbf9357f9d8594a8a743e2152e

Kotlin:       1.4.31
Groovy:       3.0.7
Ant:          Apache Ant(TM) version 1.10.9 compiled on September 27 2020
JVM:          11.0.8 (Amazon.com Inc. 11.0.8+10-LTS)
OS:           Windows 10 10.0 amd64

```

#### External tools
* sha1sum.exe from https://cygwin.com/packages/summary/perl-Digest-SHA1.html

**Comment by atakar** [comment]
Any updates regarding this? BR Atalay

**Comment by ov7a** [comment]
Sorry for the late reply.

The issue is in the backlog of the relevant team and is prioritized by them.

***

This is still reproducible in 8.14.1. It's related to the file order of the files included in the archive.

```
% unzip -l ./build/Test.zip 
Archive:  ./build/Test.zip
  Length      Date    Time    Name
---------  ---------- -----   ----
        0  02-01-1980 00:00   file1
        0  02-01-1980 00:00   file2
        0  02-01-1980 00:00   file3
        0  02-01-1980 00:00   file4
---------                     -------
        0                     4 files

% unzip -l ./build/Test2.zip
Archive:  ./build/Test2.zip
  Length      Date    Time    Name
---------  ---------- -----   ----
        0  02-01-1980 00:00   file3
        0  02-01-1980 00:00   file4
        0  02-01-1980 00:00   file1
        0  02-01-1980 00:00   file2
---------                     -------
        0                     4 files
```

Related:
* https://github.com/gradle/gradle/issues/30871

**Comment by asodja** [comment]
Hey @atakar, thanks for interest in reproducible archives. 

This is actually not a bug, we even have a test for it, see [ReproducibleArchivesIntegrationTest.groovy#L131-L173](https://github.com/gradle/gradle/blob/1d0e87137d6dc27c4f059eee2d0c86a959813c0a/subprojects/core/src/integTest/groovy/org/gradle/api/tasks/ReproducibleArchivesIntegrationTest.groovy#L131-L173). This offers users to put some files in front other files when needed. One example is, that you want MANIFEST file as first entry in a JAR, while keeping JAR reproducible.
The idea of reproducible archives is not that files are sorted in some specific order, but that the same build definition produces a byte for byte equivalent archives.

**Comment by atakar** [comment]
Hi @asodja, 
thanks for your reply. If this is an intended behavior, the Gradle task should *not* be marked `UP-TO-DATE` and skipped, once the declaration order of the `from()` calls differs. So in my opinion this is still a bug and causes confusing behavior.
If this is intentional due to Reproducible archives it has to be made sure that the Gradle`UP-TO-DATE`  and `FROM-CACHE` checks consider the declaration order and mark it as invalidated once something was changed.

**Comment by asodja** [comment]
> If this is an intended behavior, the Gradle task should not be marked UP-TO-DATE and skipped, once the declaration order of the from() calls differs.

Yes, you are right, that looks like an issue. It's not related directly to reproducible archives, but it's a general issue with AbstractArchiveTask/AbstractCopyTask and subclasses, since it doesn't track inputs properly. 
I think opening a separate issue for that would make sense, since it's a not directly related to reproducible archives and this issue focuses on reproducible archives.
